File: src/Leitor/LeitorDisciplina.java
package Leitor;

import java.util.Scanner;

import Controlador.ControladorDocente;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;

import Modelo.Disciplina;
import Modelo.Periodo;
import Modelo.Docente;

public class LeitorDisciplina extends ILeitor implements Serializable{
    
    private static final long serialVersionUID = 1L;
    private HashMap<String, Disciplina> mapa = new HashMap<>();
    private static LeitorDisciplina leitor;

    private LeitorDisciplina(){}

    public static LeitorDisciplina obterInstancia(){
        if (leitor == null){
            leitor = new LeitorDisciplina();
            return leitor;
        }
        else return leitor;
    }

    public HashMap<String,Disciplina> obterHash(){
        return mapa;
    }

    public static void setInstancia(LeitorDisciplina novo){
        leitor = novo;
    }

    public Disciplina ler(Scanner scanner, Periodo periodo, ControladorDocente controlador) throws Exception {
        String codigo = scanner.next().trim();
        
        if (busca(codigo+"-"+periodo.obterCodigo()) != null){
            throw new RuntimeException("Cadastro repetido: "+codigo+"-"+periodo.obterCodigo()+".");
        }
        String nome = scanner.next().trim();

        String login = scanner.next().trim();
        Docente docente = controlador.busca(login);
        if(docente == null){
            throw new Exception("Referência inválida: " + login + ".");
        }
        if(scanner.hasNext())
            scanner.nextLine();

        Disciplina disciplina = new Disciplina(codigo, nome, periodo, docente);
        return disciplina;
    }

    public void anexaHash(Disciplina disciplina){
        mapa.put(disciplina.obterCodigo()+"-"+disciplina.obterPeriodo().obterAno()+"/"+
            disciplina.obterPeriodo().obterSemestre(), disciplina);
    }

    public void listar(){
        System.out.println("------------------------");
		System.out.println("Disciplinas cadastradas:");
        for (String i : mapa.keySet()){
            System.out.println(mapa.get(i));
        }
    }

    public Disciplina busca(String codigo){
        if(mapa.containsKey(codigo)){
            return mapa.get(codigo);
        }
        return null;
    }

    public ArrayList<Disciplina> busca(Periodo periodo){
        ArrayList<Disciplina> lista = new ArrayList<>();
        for (String i : mapa.keySet()){
            if(mapa.get(i).obterPeriodo()==periodo)
                lista.add(mapa.get(i));
        }
        return lista;
    }

    public ArrayList<Disciplina> busca(Docente docente){
        ArrayList<Disciplina> lista = new ArrayList<>();
        for (String i : mapa.keySet()){
            if(mapa.get(i).obterDocente()==docente)
                lista.add(mapa.get(i));
        }
        return lista;
    }
}

File: cpp_version/Leitor/LeitorDisciplina.cpp
#include "LeitorDisciplina.h"

LeitorDisciplina* LeitorDisciplina::leitor = nullptr;

LeitorDisciplina::LeitorDisciplina(){}

LeitorDisciplina* LeitorDisciplina::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorDisciplina();
        return leitor;
    }
    else return leitor;
}

vector<Disciplina*> LeitorDisciplina::busca(Periodo* periodo){
    vector<Disciplina*> lista;
    for(map<string,Disciplina*>::iterator it = mapa.begin(); it != mapa.end(); ++it){
        if(it->second->obterPeriodo() == periodo)
            lista.push_back(it->second);
    }
    return lista;
}

vector<Disciplina*> LeitorDisciplina::busca(Docente* docente){
    vector<Disciplina*> lista;
    for(map<string,Disciplina*>::iterator it = mapa.begin(); it != mapa.end(); ++it){
        if(it->second->obterDocente() == docente)
            lista.push_back(it->second);
    }
    return lista;
}

Disciplina* LeitorDisciplina::busca(string codigo){
    for(map<string,Disciplina*>::iterator it = mapa.begin(); it != mapa.end(); ++it)
        if(it->first == codigo)
            return it->second;
    return nullptr;
}

Disciplina* LeitorDisciplina::ler(vector<string> vec, Periodo* periodo){
    ControladorDocente cDocente;
    string trash;
    
    string codigo = vec[1];

    if(busca(codigo + "-" + periodo->obterCodigo()) != NULL){
        throw ExcecaoCad("Cadastro repetido: "+codigo+"-"+periodo->obterCodigo()+".");
    }

    string nome = vec[2];
    string login = vec[3];

    Docente* docente = cDocente.busca(login);
    if(docente == NULL){
        throw ExcecaoRef("Referência inválida: " + login + ".");
    }

    Disciplina* disciplina = new Disciplina(codigo, nome, periodo, docente);
    return disciplina;
}

void LeitorDisciplina::anexaHash(Disciplina* disciplina){
    mapa[disciplina->obterCodigo()+"-"+to_string(disciplina->obterPeriodo()->obeterAno())+"/"+
        disciplina->obterPeriodo()->obterSemestre()] = disciplina;
}

vector<Disciplina*> LeitorDisciplina::obterDisciplinas(){
    vector<Disciplina*> lista;
    for(map<string,Disciplina*>::iterator it = mapa.begin(); it != mapa.end(); ++it)
        lista.push_back(it->second);
    return lista;
}